libdeno/binding.cc

```diff
@@ -349,34 +343,145 @@ void Shared(v8::Local<v8::Name> property,
+  // it's quite unique enough for the purposes of O(1000) modules.  Should check
+  // that map isn't overwritting previous values so we at least get a debuggable
+  // error if that ever occurs.
+  int id = module->GetIdentityHash();
+  module_filename_map_[id] = filename;
+  module_map_.emplace(std::piecewise_construct, std::make_tuple(filename),
+                      std::make_tuple(isolate_, module));
+}
+
+v8::MaybeLocal<v8::Module> CompileModule(v8::Local<v8::Context> context,
+                                         const char* js_filename,
+                                         v8::Local<v8::String> source_text) {
   auto* isolate = context->GetIsolate();
+
   v8::Isolate::Scope isolate_scope(isolate);
-  v8::HandleScope handle_scope(isolate);
+  v8::EscapableHandleScope handle_scope(isolate);
+  v8::Context::Scope context_scope(context);
+
+  auto origin = ModuleOrigin(v8_str(js_filename, true), isolate);
+  v8::ScriptCompiler::Source source(source_text, origin);
+
+  auto maybe_module = v8::ScriptCompiler::CompileModule(isolate, &source);
+
+  if (!maybe_module.IsEmpty()) {
+    auto module = maybe_module.ToLocalChecked();
+    CHECK_EQ(v8::Module::kUninstantiated, module->GetStatus());
+    DenoIsolate* d = FromIsolate(isolate);
+    d->RegisterModule(js_filename, module);
+  }
 
+  return handle_scope.EscapeMaybe(maybe_module);
+}
+
+v8::MaybeLocal<v8::Module> ResolveCallback(v8::Local<v8::Context> context,
+                                           v8::Local<v8::String> specifier,
+                                           v8::Local<v8::Module> referrer) {
+  auto* isolate = context->GetIsolate();
+  DenoIsolate* d = FromIsolate(isolate);
+
+  v8::Isolate::Scope isolate_scope(isolate);
+  v8::EscapableHandleScope handle_scope(isolate);
   v8::Context::Scope context_scope(context);
 
-  v8::TryCatch try_catch(isolate);
+  int ref_id = referrer->GetIdentityHash();
+  std::string referrer_filename = d->module_filename_map_[ref_id];
 
-  auto name = v8_str(js_filename);
+  v8::String::Utf8Value specifier_(isolate, specifier);
+  const char* specifier_c = ToCString(specifier_);
 
-  v8::ScriptOrigin origin(name);
+  CHECK_NE(d->resolve_cb_, nullptr);
+  d->resolve_cb_(d->user_data_, specifier_c, referrer_filename.c_str());
 
-  auto script = v8::Script::Compile(context, source, &origin);
+  if (d->resolve_module_.IsEmpty()) {
+    // Resolution Error.
+    isolate->ThrowException(v8_str("module resolution error"));
+    return v8::MaybeLocal<v8::Module>();
+  } else {
+    auto module = d->resolve_module_.Get(isolate);
+    d->resolve_module_.Reset();
+    return handle_scope.Escape(module);
+  }
+}
 
-  if (script.IsEmpty()) {
+void DenoIsolate::ResolveOk(const char* filename, const char* source) {
+  CHECK(resolve_module_.IsEmpty());
+  auto count = module_map_.count(filename);
+  if (count == 1) {
+    auto module = module_map_[filename].Get(isolate_);
+    resolve_module_.Reset(isolate_, module);
```

Comment:
Does this mean that there are chances where we might fetch a source multiple times, but discard all the fetched source code and only using the first fetched version?

(As a side question, does V8 `InstantiateModule` guarantees that the same module would not be running in the same context twice?)

Reply:
I believe (tho I'm not sure, I haven't tested it) that the `module_map_` will properly prevent that from happening. The `resolve_module_.Reset()` here is only to pass around a reference during resolution.